rbd: perform resize of file system for static volume

Madhu-1 opened the pull request:
If the rbd image is already formatted and mounted on the node, If the user resizes the image on the ceph cluster and deletes and recreates the kubernetes objects, the new size will not be visible on the node.

With this change, it will check the size of the node and also the rbd image size on the ceph cluster. If the size is not matching it will do the file system resize on the node.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

closes #2552

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -563,6 +595,27 @@ func (ns *NodeServer) createStageMountPoint(ctx context.Context, mountPath strin
 	return nil
 }
 
+func getDeviceSize(ctx context.Context, devPath string) (int64, error) {
```
how about rewriting this and using the functions from `vendor/k8s.io/kubernetes/pkg/volume/metrics_block.go`:
```go
mp := NewMetricsBlock(devPath)
m, err := mp.GetMetrics()
return metrics.Capacity, nil
```

Madhu-1 replied:
Need to extract the size in bytes.  added now. let see if it works.

nixpanic replied:
Yes, of course. I was just too lazy to check how to get the bytes from the capacity :sheep:

nixpanic reviewed (changes requested):
Looks good, but the commit message is a little unclear, please phrase it a little better. Maybe also mention what actions need to be taken to resize the image in the docs?

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -563,6 +596,19 @@ func (ns *NodeServer) createStageMountPoint(ctx context.Context, mountPath strin
+	mp := volume.NewMetricsBlock(devPath)
+	m, err := mp.GetMetrics()
+	if err != nil {
+		err = fmt.Errorf("failed to get metrics %w", err)
```
nit: ideally add a `: ` just before `%w`

Madhu-1 replied:
Thanks added now.

nixpanic reviewed (dismissed):
thanks for the update and improved docs!

Yuggupta27 commented on docs/static-pvc.md:
````diff
@@ -124,6 +125,26 @@ $ kubectl create -f fs-static-pvc.yaml
+rbd resize static-image --size=2048 --pool=replicapool
+```
+
+Once the rbd image is resized in the ceph cluster. update the PV size and PVC
+size to match the rbd image size to match the size of the rbd image.
````
```suggestion
Once the rbd image is resized in the ceph cluster, update the PV size and PVC
size to match the size of the rbd image.
```

humblec commented on docs/static-pvc.md:
```diff
@@ -124,6 +125,26 @@ $ kubectl create -f fs-static-pvc.yaml
 
+### Resize RBD image
+
+Lets resize the RBD image in ceph cluster
```
Let us

yati1998 commented:
ci/centos/mini-e2e/k8s-1.20  fails with error: `Oct  4 16:06:51.650: FAIL: failed to validate rbd static pv with error timed out waiting for the condition`

Madhu-1 commented:
am able to resize the filesystem in manual testing. Added some useful log to check what is the issue.

humblec commented on e2e/staticpvc.go:
```diff
@@ -189,6 +201,39 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkI
+	pvc *v1.PersistentVolumeClaim,
+	rbdImageName string) error {
+	// resize rbd image
+	size := "8Gi"
```
may be declare or use a var for this size?

Madhu-1 replied:
isn't it already a short variable declaration? did you mean something else?

Madhu-1 replied:
declared a new global variable

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -563,6 +602,19 @@ func (ns *NodeServer) createStageMountPoint(ctx context.Context, mountPath strin
 }
 
+func getDeviceSize(ctx context.Context, devPath string) (int64, error) {
+	mp := volume.NewMetricsBlock(devPath)
```
okay, this does not work as a check for resizing. because if the rbd image is 8Gib when we check the size of the device it always returns the size of the device (8Gib), not the filesystem size which we can use for comparison.

@nixpanic @humblec @pkalever any other idea which we can use for size comparison to doing resize? or another option is to blindly resize which is resize the filesystem if it's required or else it will return success.

Madhu-1 replied:
Discard this one. The latest changes should fix it.

Madhu-1 committed:
rbd: use NewMetricsBlock to get size

instead of lsblk command use NewMetricsBlock
function from the kubernetes package to get
the size.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: move flattening to helper function

in NodeStage operation we are flattening
the image to support mounting on the older
clients. this commits moves it to a helper
function to reduce code complexity.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: perform resize of file system for static volume

For static volume, the user will manually mounts
already existing image as a volume to the application
pods. As its a rbd Image, if the PVC is of type
fileSystem the image will be mapped, formatted
and mounted on the node,
If the user resizes the image on the ceph cluster.
User cannot not automatically resize the filesystem
created on the rbd image. Even if deletes and
recreates the kubernetes objects, the new size
will not be visible on the node.

With this changes During the NodeStageVolumeRequest
the nodeplugin will check the size of the mapped rbd
image on the node using the devicePath. and also
the rbd image size on the ceph cluster.

If the size is not matching it will do the file
system resize on the node as part of the
NodeStageVolumeRequest RPC call.

The user need to do below operation to see new size
* Resize the rbd image in ceph cluster
* Scale down all the application pods using the static
PVC.
* Make sure no application pods which are using the
static PVC is running on a node.
* Scale up all the application pods.

Validate the new size in application pod mounted
volume.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

yati1998 reviewed (commented):
LGTM

Madhu-1 commented:
@Yuggupta27 You have requested changes on this PR. PTAL
